entity-service-client/src/main/java/org/hypertrace/entity/query/service/client/EntityLabelsClient.java

```diff
@@ -0,0 +1,27 @@
+package org.hypertrace.entity.query.service.client;
+
+import java.util.List;
+import java.util.Map;
+
+public interface EntityLabelsClient {
+  List<String> getEntityLabelsForEntity(String idColumnName,
```

Comment:
I thought you wanted things to be async?!

Regardless, I find this api a bit confusing. Why does the caller have to provide column names? And why are we explicitly unwrapping things like tenantIds and headers - that defeats the purpose of using a context which automatically propagates them.

Also, what's the reasoning to make this explicit to labels vs entities in general?

Reply:
RequestContext is defined in gateway-service that's why we have to pass in tenantId(needed for the cache key) and headers(needed by the actual entity-service client).

The caller provides column names since they are needed to construct the query to entity service in case of a cache miss. And yes, this cache is special for entity labels. In order to be able to make queries to entity service incase of a cache miss, we need all these arguments. Something more generic would need more thought and time :)

Also while I would like this to be async, it does not help at the moment since I need to know the entityIds whose labels I need to fetch. So might as well tackle that while making all API calls async.

Reply:
>  In order to be able to make queries to entity service incase of a cache miss, we need all these arguments. 

Entity service exposes column names?!

> Something more generic would need more thought and time :)

That's fair - I'm not necessarily asking for changes yet, just trying to understand what led us here

I guess a more macro level question is - why is any of this needed? Gateway today has an entity fetcher (no idea if it caches, but presumably it could be added). What's special about labels that requires there own client when other entity-based attributes do not?

Reply:
> Entity service exposes column names?!

No. They are based on attribute service. So they correspond to the attribute metadata ids. It's expected that entity-service respects that but we do pass them into the query as strings.

> That's fair - I'm not necessarily asking for changes yet, just trying to understand what led us here
> 
> I guess a more macro level question is - why is any of this needed? Gateway today has an entity fetcher (no idea if it caches, but presumably it could be added). What's special about labels that requires there own client when other entity-based attributes do not?

Oh. 2 main motivations:
- Performance considerations; avoiding queries into 2 data sources like we were doing before since the labels will be used to annotate the entities - Service and API - wherever they appear on the UI.
- Tied to the bullet above, since we will be associating entity labels as foreign columns in other assets/scopes, and we are not writing these to Pinot, then we need a performant way to append labels to results of these other scopes.

So, this is a quick cache to get the feature out there and maintain some of the performance numbers that we have. @sanjay-nagaraj and I were talking about exploring a real cache service like redis.

Reply:
> No. They are based on attribute service. So they correspond to the attribute metadata ids. It's expected that entity-service respects that but we do pass them into the query as strings.

If I understood correctly, we may want to change the naming to something like idAttributeKey or something (or ID - depending on if scope is included in the value). Ideally, we should be able to omit these and have entity service look that information up internally though. I guess that requires some widespread changes to the API since it's currently exposed like query service, as a low level sql-style api.

> Performance considerations; avoiding queries into 2 data sources like we were doing before since the labels will be used to annotate the entities - Service and API - wherever they appear on the UI.

Agree, that makes sense - although I think a general entity cache would serve that role just as well and slot in easier.

> Tied to the bullet above, since we will be associating entity labels as foreign columns in other assets/scopes, and we are not writing these to Pinot, then we need a performant way to append labels to results of these other scopes.

This one scares me - is the foreign scope need immediate? I would want to think this one through because we're basically implementing a specialized join for this one attribute. Everywhere else in the platform, we do that at write time - knowing that we want to move it to read time through something like presto eventually. I would agree that doing this generically is out of scope, but I thought we had said it was acceptable and we didn't need that functionality right now? If it is required, I would push to implement it in the same manner as every other join rather than introduce a whole new concept that doesn't solve the problem for other joins.

Reply:
> If I understood correctly, we may want to change the naming to something like idAttributeKey or something (or ID - depending on if scope is included in the value). Ideally, we should be able to omit these and have entity service look that information up internally though. I guess that requires some widespread changes to the API since it's currently exposed like query service, as a low level sql-style api.

Actually I will backtrack, sort of. Since entity-service has no knowledge of attributes in attribute service, then these are the attribute names in the doc store config defined for entities in mongo

```
entity.service.attributeMap = [
  {
    "scope": "API",
    "name": "API.id",
    "subDocPath": "entityId"
  },
  {
    "scope": "API",
    "name": "API.name",
    "subDocPath": "entityName"
  },
  {
    "scope": "API",
    "name": "API.apiType",
    "subDocPath": "attributes.API_TYPE"
  },
...
]
``` 
> > Tied to the bullet above, since we will be associating entity labels as foreign columns in other assets/scopes, and we are not writing these to Pinot, then we need a performant way to append labels to results of these other scopes.
> 
> This one scares me - is the foreign scope need immediate? I would want to think this one through because we're basically implementing a specialized join for this one attribute. Everywhere else in the platform, we do that at write time - knowing that we want to move it to read time through something like presto eventually. I would agree that doing this generically is out of scope, but I thought we had said it was acceptable and we didn't need that functionality right now? If it is required, I would push to implement it in the same manner as every other join rather than introduce a whole new concept that doesn't solve the problem for other joins.

The foreign scope need is to be able to do queries on `API_TRACE.apiLabels` for example where labels are not written to Pinot.

We had issues with write time joins since unlike other attributes this one, being an entity attribute, not an attribute on the events associated with the entity can mutate and Pinot does not allow for mutations. Also, the logic to do the joins is the same whether the foreign scope equals primary scope or is different. i.e querying for `API.labels` and `API_TRACE.apiLabels` will go through the same path.

Also, unless I'm confused I thought we agreed to do it this way and avoid writing the labels to Pinot? Writing to Pinot would not be ideal when querying for entities. Consider the services list query for example. When querying from Pinot we group by service id and service name in order to be able to get the aggregated metrics for all the services together with the name and id in one query. If labels were to be involved, since they are an attribute, the results would not be what we want if we grouped by that column - since it's a string array. So we fallback to querying EDS for labels to avoid this weird group by and if doing it for the primary scopes and foreign scopes is the same logic then I'd rather do it this way.
CC @buchi-busireddy @sanjay-nagaraj

Reply:
> The foreign scope need is to be able to do queries on API_TRACE.apiLabels for example where labels are not written to Pinot.

I didn't realize querying traces by label was a requirement for the first cut - My impression, perhaps incorrect, was that we first just wanted to be able to see tags on entities and update them.

> We had issues with write time joins since unlike other attributes this one, being an entity attribute, not an attribute on the events associated with the entity can mutate and Pinot does not allow for mutations. 

Didn't we need to do this with isExternal for something around domains? I thought I recalled this being done before. Anyway, I agree with the points about the implications of writing to pinot - I thought we were avoiding that too by not supporting those types of functionality right now, rather than implementing the join ourselves.

Reply:
Anyway, since it's already implemented the discussion here was more about limiting the scope so is not super important. Will take a look at what's going on in gateway and how that's used - as long as there's not lots of custom code around labels, we're good.

Reply:
> > The foreign scope need is to be able to do queries on API_TRACE.apiLabels for example where labels are not written to Pinot.
> 
> I didn't realize querying traces by label was a requirement for the first cut - My impression, perhaps incorrect, was that we first just wanted to be able to see tags on entities and update them.

Oh you can do that right now once @arjunlalb puts in the UI changes. Though as it is, if you want to avoid queries to both QS and EDS for a single entity query in gw service, you have to structure your queries in the UI so the labels one is not part of one requesting for metrics. This works on API and Service overview pages.

If you want the labels displayed on the entity list pages, then it gets interesting. I guess you can still make 2 requests but doing the joining at the UI gets tricky. So that's where a cache in gateway service comes in.

> > We had issues with write time joins since unlike other attributes this one, being an entity attribute, not an attribute on the events associated with the entity can mutate and Pinot does not allow for mutations.
> 
> Didn't we need to do this with isExternal for something around domains? I thought I recalled this being done before. Anyway, I agree with the points about the implications of writing to pinot - I thought we were avoiding that too by not supporting those types of functionality right now, rather than implementing the join ourselves.

`isExternal` and `apiDiscoveryState are not big problems since they are simple strings. With labels it's tricky since it's a string array. Also it can mutate.

Reply:
> Oh you can do that right now once @arjunlalb puts in the UI changes. Though as it is, if you want to avoid queries to both QS and EDS for a single entity query in gw service, you have to structure your queries in the UI so the labels one is not part of one requesting for metrics. This works on API and Service overview pages.

As far as I know, the only places we're querying labels are on entities queries - either for the entity list or overviews. The UI, at this point (@arjunlalb correct me if I'm wrong) isn't working with labels at the trace level, and the new attributes you added in helm-values are only on entities too.

As to the query structure, is there any benefit of sending two queries on the UI side vs sending on query and making the gateway fire two queries?

I do get the benefit of the cache, it just seems like it should support entities more broadly vs a specific cache for labels - because that implies gateway service requires specific logic for labels too.